### PR TITLE
Readyapi4j cucumber

### PR DESCRIPTION
**Pull request by olensmar**
this PR merges the standalone readyapi/testserver-cucumber repo into this - all content is in the modules/cucumber module. Global README udpates still needed for links, etc

Updated Cucumber Docker image is available at https://hub.docker.com/r/smartbear/readyapi4j-cucumber/

### CODE REVIEW

File: modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/builders/Assertions.java [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+import com.smartbear.readyapi.client.model.ValidHttpStatusCodesAssertion;
+import com.smartbear.readyapi.client.model.XPathContainsAssertion;
+
+import java.util.List;
+
+/**
+ * Utility class with static method for building various types of assertions
+ */
+
+public class Assertions {
```
Comment (prakashmrec) [inline modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/builders/Assertions.java]: Some of the methods in this class are identical to methods in Assertions class in readyapi4j-core module. I think either we can remove the duplicates or delegate to Assertions class in readyapi4j-core for duplicates.

File: modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+/**
+ * Executes a TestServer Recipe class using the TestServer endpoint specified
+ * in the testserver.endpoint system property (defaults to the public
+ * TestServer instance).
+ */
+
+public class CucumberRecipeExecutor {
+
+    private final static Logger LOG = LoggerFactory.getLogger(CucumberRecipeExecutor.class);
+    private static final String TESTSERVER_ENDPOINT = "testserver.endpoint";
```
Comment (prakashmrec) [inline modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java]: These constants are not used anywhere in this class ?

File: modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+    private static final String TESTSERVER_USER = "testserver.user";
+    private static final String TESTSERVER_PASSWORD = "testserver.password";
+
+    private RecipeExecutor executor;
+    private boolean async = false;
+
+    public CucumberRecipeExecutor() {
+        executor = RecipeExecutorBuilder.buildDefault();
```
Comment (prakashmrec) [inline modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java]: Does this mean that it will be local execution if the system properties are not present for TestServer ?

File: modules/cucumber/modules/core/pom.xml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>4.12</version>
+        </dependency>
+
+        <dependency>
+            <groupId>com.smartbear.readyapi</groupId>
+            <artifactId>readyapi4j-facade</artifactId>
+            <version>2.0.0-SNAPSHOT</version>
```
Comment (prakashmrec) [inline modules/cucumber/modules/core/pom.xml]: This should be ${project.version}

File: modules/cucumber/modules/runner/src/docker/resources/soapui-log4j.xml [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+        <param name="Threshold" value="INFO"/>
+        <layout class="org.apache.log4j.PatternLayout">
+            <!-- The default pattern: Date Priority [Category] Message\n -->
+            <param name="ConversionPattern" value="%d{ABSOLUTE} %-5p [%c{1}] %m%n"/>
+        </layout>
+    </appender>
+
+    <appender name="FILE" class="org.apache.log4j.RollingFileAppender">
+        <errorHandler class="org.apache.log4j.helpers.OnlyOnceErrorHandler"/>
+        <param name="File" value="${soapui.logroot}soapui.log"/>
```
Comment (prakashmrec) [inline modules/cucumber/modules/runner/src/docker/resources/soapui-log4j.xml]: What will be ${soapui.logroot} in case of docker, the user have to provide the value or is it mapped by default?

File: modules/cucumber/modules/runner/src/docker/resources/soapui-log4j.xml [hunk 6]
```diff
@@ -0,0 +1,105 @@
...
+
+    <logger name="groovy.log">
+        <level value="INFO"/>
+        <appender-ref ref="GLOBAL_GROOVY_LOG"/>
+    </logger>
+    <!-- =============== -->
+    <!-- Loggers         -->
+    <!-- =============== -->
+
+    <logger name="com.eviware.soapui">
```
Comment (prakashmrec) [inline modules/cucumber/modules/runner/src/docker/resources/soapui-log4j.xml]: Most of the appenders in this log4j config are for local execution. Do we really want to print SoapUI logs when executing recipes? I think we should add appender for cucumber step defs and add logging messages in those classes.

**Comment by prakashmrec** [comment]
Conflicts :(

File: modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java (hunk 3 above)
Comment (olensmar) [inline modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java]: yes!

File: modules/cucumber/modules/runner/src/docker/resources/soapui-log4j.xml (hunk 6 above)
Comment (olensmar) [inline modules/cucumber/modules/runner/src/docker/resources/soapui-log4j.xml]: makes sense - will look into it

File: modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/builders/Assertions.java (hunk 1 above)
Comment (olensmar) [inline modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/builders/Assertions.java]: thanks - will move!

File: modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java (hunk 2 above)
Comment (olensmar) [inline modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java]: removing...

File: modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java (hunk 3 above)
Comment (spindelmanne) [inline modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java]: Good! :-)

**Comment by prakashmrec** [comment]
Looks good to me now, @spindelmanne ?

File: modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeBuilder.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+import com.smartbear.readyapi4j.support.AssertionUtils;
+import cucumber.api.Scenario;
+import cucumber.api.java.After;
+import cucumber.runtime.java.guice.ScenarioScoped;
+
+import javax.inject.Inject;
+import java.util.List;
+
+/**
+ * Class that builds and runs the recipe to be executed
```
Comment (olensmar) [inline modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeBuilder.java]: this class "merges" readyapi4j into the cucumber framework via the @After and @ScenarioScoped annotations - perhaps it should be named accordingly/differently?

File: modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+            }
+
+            String[] pathSegments = scenario.getId().split(";");
+            File scenarioFolder = folder;
+            int fileIndex = 0;
+
+            if( pathSegments.length > 1 ) {
+                scenarioFolder = new File(folder, pathSegments[0]);
+                if (scenarioFolder.exists() || !scenarioFolder.isDirectory()) {
+                    scenarioFolder.mkdirs();
```
Comment (olensmar) [inline modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java]: perhaps - what should happen if it fails? exception? ignore?

File: modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+     *
+     * @param testRecipe the test recipe to log
+     * @param scenario the associated Cucumber scenario
+     * @param logFolder the root folder for generated folders and files
+     */
+
+    protected void logScenarioToFile(TestRecipe testRecipe, Scenario scenario, String logFolder) {
+        try {
+            File folder = new File( logFolder );
+            if( !folder.exists() || !folder.isDirectory()){
```
Comment (olensmar) [inline modules/cucumber/modules/core/src/main/java/com/smartbear/readyapi/readyapi4j/cucumber/CucumberRecipeExecutor.java]: suggest we do this globally in a separate pr before release

**Comment by olensmar** [comment]
@spindelmanne please see updates/comments

**Comment by spindelmanne** [comment]
OK
